sdks/python/apache_beam/pipeline.py

```diff
@@ -789,6 +797,10 @@ def _infer_result_type(
       # Any remaining type variables have no bindings higher than this scope.
       result_pcollection.element_type = typehints.bind_type_variables(
           result_element_type, {'*': typehints.Any})
+      # TODO: Where should we actually do this?
+      if isinstance(transform, ParDo):
+        transform.infer_batch_converter(input_element_type, input_batch_type)
+        result_pcollection.batch_type = transform.infer_output_batch_type()
```

Comment:
It's insufficient to rely on the next DoFn to resolve this batch converter.

What if the output is a Flatten or there's a fusion break (to test this add a side input between a and b).